orbotservice/src/main/java/org/torproject/android/service/circumvention/SnowflakeProxyWrapper.kt

```diff
@@ -22,37 +32,59 @@ class SnowflakeProxyWrapper(private val context: Context) {
         if (proxy != null) return
         if (Prefs.limitSnowflakeProxyingWifi() && !hasWifi) return
         if (Prefs.limitSnowflakeProxyingCharging() && !hasPower) return
-        proxy = SnowflakeProxy()
-        val stunServers = BuiltInBridges.getInstance(context)?.snowflake?.firstOrNull()?.ice
-            ?.split(",".toRegex())?.dropLastWhile { it.isEmpty() }?.toTypedArray() ?: emptyArray()
-        val stunUrl = stunServers[SecureRandom().nextInt(stunServers.size)]
-
-        proxy = SnowflakeProxy()
-        with(proxy!!) {
-            brokerUrl = OrbotService.getCdnFront("snowflake-target-direct")
-            capacity = 1L
-            pollInterval = 120L
-            stunServer = stunUrl
-            relayUrl = OrbotService.getCdnFront("snowflake-relay-url")
-            natProbeUrl = OrbotService.getCdnFront("snowflake-nat-probe")
-            clientConnected = SnowflakeClientConnected { onConnected() }
-            start()
-        }
 
-        if (Prefs.showSnowflakeProxyMessage()) {
-            val message = context.getString(R.string.log_notice_snowflake_proxy_enabled)
-            Handler(context.mainLooper).post {
-                context.applicationContext.showToast(message)
+        CoroutineScope(Dispatchers.IO).launch {
+            val start = Random.nextInt(49152, 65536 - 2)
+
+            for (port in (start..start + 2)) {
+                if (UPnP.openPortUDP(port, OrbotConstants.TAG)) {
+                    mappedPorts.add(port)
+                }
+            }
+
+            if (mappedPorts.size < 3) {
+                releaseMappedPorts()
+            }
+
+            val stunServers = BuiltInBridges.getInstance(context)?.snowflake?.firstOrNull()?.ice
+                ?.split(",".toRegex())?.dropLastWhile { it.isEmpty() }?.toTypedArray() ?: emptyArray()
+            val stunUrl = stunServers[SecureRandom().nextInt(stunServers.size)]
+
+            proxy = SnowflakeProxy()
+            with(proxy!!) {
+                brokerUrl = OrbotService.getCdnFront("snowflake-target-direct")
+                capacity = 1L
+                pollInterval = 120L
+                stunServer = stunUrl
+                relayUrl = OrbotService.getCdnFront("snowflake-relay-url")
+                natProbeUrl = OrbotService.getCdnFront("snowflake-nat-probe")
+                clientConnected = SnowflakeClientConnected { onConnected() }
+
+// TODO: Activate, when new IPtProxy is available.
+//            ephemeralMinPort = mappedPorts.firstOrNull()
```

Comment:
will IPtProxy be ok with getting null as an argument, just because if mappedPorts is < 3 the ports are released and then mappedPorts is reassigned to an empty list, meaning that `ephemeralMinPort` and `ephemeralMaxPort would in fact be assigned to null here

Reply:
Yes, that's why this is done. Setting max/min ports to null is like setting it not at all.
Needs another comment, I guess.